lib/charms/grafana_k8s/v1/grafana_source.py

```diff
@@ -0,0 +1,336 @@
+                providers. The keys of the dictionary are :string:
+                names of grafana source service providers. Typically,
+                this is `grafana-source`. The values of the dictionary
+                are corresponding minimal acceptable semantic versions
+                for the service.
+            multi: an optional (default `False`) flag to indicate if
+                this object should support interacting with multiple
+                service providers.
+
+        """
+        super().__init__(charm, name, consumes, multi)
+
+        self.charm = charm
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
+        events = self.charm.on[name]
+        self.framework.observe(events.relation_joined, self._set_sources)
+
+    def add_source(self, address: str, port: str,
+                   source_type: str = "prometheus", source_name: Optional[str] = None,
+                   rel_id: Optional[int] = None) -> None:
+        """Add an additional source to the Grafana source service.
+
+        Args:
+            address: a :str: which gives address of the target
+            port: a :str: which gives the port of the target
+            source_type: a :str: indicating the source type. Defaults to "prometheus"
+            source_name: a :str: specifying the source name. Otherwise this is calculated
+            rel_id: an optional integer specifying the relation ID
+                for the grafana source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+        rel_id = rel_id if rel_id is not None else rel.id
+
+        unique_source_name = "juju_{}_{}_{}".format(
+            self.charm.model.name,
+            self.charm.app.name,
+            rel_id
+        )
+        if not source_name or self._source_name_in_use(source_name):
+            logger.warning(
+                "'source-name' not specified' or provided name is already in use. "
+                "Using safe default: {}.".format(unique_source_name)
+            )
+            source_name = unique_source_name
+
+        data = {
+            "address": address,
+            "port": port,
+            "source-type": source_type,
+            "source-name": source_name
+        }
+
+        self._update_sources(data, rel_id)
+
+    def remove_source(self, rel_id=None) -> None:
+        """Removes a source relation.
+
+        Args:
+            rel_id: an optional integer specifying the relation ID
+                for the grafana-k8s source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
```

Comment:
will be None if no relations in place

Reply:
In which case the consumer will never be ready.

Reply:
how about after remove-relation?
you mean you have `.ready` and `.unready` guards?

Reply:
This code has already been merged into `grafana-operator`. With approvals. Including you.

There is no reason we would expect the user of a consumer library to try to invoke it after the relationship is broken. And if they do, an exception is _expected_ and good practice Python. Exceptions are not nearly as forbidden in Python as other languages

Reply:
Apologies for entering into nitpick land here, but:
The exception would be coming from your new code, not library code, i.e. you'd be introducing a new exception which is not related to the problem: `AttributeError: 'NoneType' object has no attribute 'id'`, which seems like a coding mistake rather than exception propagation.

"There is no reason we would expect the user [...] to try to invoke it" is acceptable though.

Reply:
Correct (except this is a consumer library, which is why I called it library bode). "Something very bad happened and your expected workflow did not complete. You need to take a look." is the role of exceptions being thrown. It is the `5xx` return code of "you did something wrong" in Python if it were HTTP, rather than the `4xx` `logging.error` of "you did everything right, but something bad happened when we tried to talk to that webserver/synchronize data".